devtools: Cannot find module puppeteer-core/node_modules/debug

**Environment:**
 - **WebdriverIO version:** 6.0.5
 - **Mode:** WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:** sync
 - **Node.js version:** 12.16.1
 - **NPM version:** 6.14.4
 - **Browser name and version:** any
 - **Platform name and version:** any

**Config of WebdriverIO**
Any.

**Describe the bug**
Test runner crashes, see logs.
Possible cause: 
https://github.com/webdriverio/webdriverio/pull/5007/files#diff-ac866ff0354e029d2ba18bc72e991988R19
Locking puppeteer-core to 2.1.1 with npm shrinkwrap does not help.

**To Reproduce**
???
Upgrade to v6.

**Expected behavior**
Test runner should not not crash.

**Log**
```js
(node:10834) UnhandledPromiseRejectionWarning: Error: Cannot find module '/Users/[REDACTED]/Repos/wdio-v6/node_modules/puppeteer-core/node_modules/debug'
Require stack:
- /Users/[REDACTED]/Repos/wdio-v6/node_modules/devtools/build/index.js
- /Users/[REDACTED]/Repos/wdio-v6/node_modules/webdriverio/build/utils.js
- /Users/[REDACTED]/Repos/wdio-v6/node_modules/webdriverio/build/multiremote.js
- /Users/[REDACTED]/Repos/wdio-v6/node_modules/webdriverio/build/index.js
- /Users/[REDACTED]/Repos/wdio-v6/node_modules/@wdio/cli/build/commands/repl.js
- /Users/[REDACTED]/Repos/wdio-v6/node_modules/yargs/index.js
- /Users/[REDACTED]/Repos/wdio-v6/node_modules/@wdio/cli/build/index.js
- /Users/[REDACTED]/Repos/wdio-v6/node_modules/@wdio/cli/bin/wdio.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)
    at Function.Module._load (internal/modules/cjs/loader.js:864:27)
    at Module.require (internal/modules/cjs/loader.js:1044:19)
    at require (internal/modules/cjs/helpers.js:77:18)
    at Object.<anonymous> (/Users/[REDACTED]/Repos/wdio-v6/node_modules/devtools/build/index.js:47:1)
    at Module._compile (internal/modules/cjs/loader.js:1158:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)
    at Module.load (internal/modules/cjs/loader.js:1002:32)
    at Function.Module._load (internal/modules/cjs/loader.js:901:14)
    at Module.require (internal/modules/cjs/loader.js:1044:19)

```

--- a/docs/Babel.md
+++ b/docs/Babel.md
@@ -37,7 +37,8 @@ exports.config = {
     // ...
     mochaOpts: {
         ui: 'bdd',
-        require: ['@babel/register', './test/helpers/common.js']
+        require: ['@babel/register', './test/helpers/common.js'],
+        // ...
     },
     // ...
 }
@@ -47,8 +48,10 @@ exports.config = {
 // wdio.conf.js
 exports.config = {
     // ...
-    before: function() {
-        require('@babel/register')
+    jasmineNodeOpts: {
+        // Jasmine default timeout
+	    helpers: [require.resolve('@babel/register')],
+        // ...
     },
     // ...
 }
@@ -60,7 +63,8 @@ exports.config = {
     // ...
     cucumberOpts: {
         requireModule: ['@babel/register'],
-        require: ['./test/helpers/common.js']
+        require: ['./test/helpers/common.js'],
+        // ...
     },
     // ...
 }

--- a/packages/devtools/src/constants.js
+++ b/packages/devtools/src/constants.js
@@ -84,6 +84,8 @@ export const SUPPORTED_SELECTOR_STRATEGIES = ['css selector', 'tag name', 'xpath
 export const SERIALIZE_PROPERTY = 'data-devtoolsdriver-fetchedElement'
 export const SERIALIZE_FLAG = '__executeElement'
 
+export const PPTR_LOG_PREFIX = 'puppeteer:protocol'
+
 export const ERROR_MESSAGES = {
     staleElement: {
         name: 'stale element reference',

--- a/packages/devtools/src/index.js
+++ b/packages/devtools/src/index.js
@@ -1,3 +1,4 @@
+import fs from 'fs'
 import os from 'os'
 import path from 'path'
 import { v4 as uuidv4 } from 'uuid'
@@ -16,10 +17,19 @@ const log = logger('devtools:puppeteer')
  * log puppeteer messages
  */
 const PREFIX = 'puppeteer:protocol'
-const puppeteerDebugPkg = path.resolve(
+let puppeteerDebugPkg = path.resolve(
     path.dirname(require.resolve('puppeteer-core')),
     'node_modules',
     'debug')
+
+/**
+ * check if Puppeteer has its own version of debug, if not use the
+ * one that is installed for all packages
+ */
+if (!fs.existsSync(puppeteerDebugPkg)) {
+    puppeteerDebugPkg = require.resolve('debug')
+}
+
 /* istanbul ignore next */
 require(puppeteerDebugPkg).log = (msg) => {
     if (msg.includes('puppeteer:protocol')) {

--- a/packages/devtools/src/index.js
+++ b/packages/devtools/src/index.js
@@ -1,6 +1,4 @@
-import fs from 'fs'
 import os from 'os'
-import path from 'path'
 import { v4 as uuidv4 } from 'uuid'
 import logger from '@wdio/logger'
 import { webdriverMonad, devtoolsEnvironmentDetector } from '@wdio/utils'
@@ -9,34 +7,14 @@ import { validateConfig } from '@wdio/config'
 import DevToolsDriver from './devtoolsdriver'
 import launch from './launcher'
 import { DEFAULTS, SUPPORTED_BROWSER } from './constants'
-import { getPrototype } from './utils'
+import { getPrototype, patchDebug } from './utils'
 
 const log = logger('devtools:puppeteer')
 
 /**
- * log puppeteer messages
+ * patch debug package to log Puppeteer CDP messages
  */
-const PREFIX = 'puppeteer:protocol'
-let puppeteerDebugPkg = path.resolve(
-    path.dirname(require.resolve('puppeteer-core')),
-    'node_modules',
-    'debug')
-
-/**
- * check if Puppeteer has its own version of debug, if not use the
- * one that is installed for all packages
- */
-if (!fs.existsSync(puppeteerDebugPkg)) {
-    puppeteerDebugPkg = require.resolve('debug')
-}
-
-/* istanbul ignore next */
-require(puppeteerDebugPkg).log = (msg) => {
-    if (msg.includes('puppeteer:protocol')) {
-        msg = msg.slice(msg.indexOf(PREFIX) + PREFIX.length).trim()
-    }
-    log.debug(msg)
-}
+patchDebug(log)
 
 export const sessionMap = new Map()
 

--- a/packages/devtools/src/utils.js
+++ b/packages/devtools/src/utils.js
@@ -1,11 +1,12 @@
 import fs from 'fs'
+import path from 'path'
 import { execFileSync } from 'child_process'
 import logger from '@wdio/logger'
 import { commandCallStructure, isValidParameter, getArgumentType } from '@wdio/utils'
 import { WebDriverProtocol } from '@wdio/protocols'
 
 import cleanUp from './scripts/cleanUpSerializationSelector'
-import { ELEMENT_KEY, SERIALIZE_PROPERTY, SERIALIZE_FLAG, ERROR_MESSAGES } from './constants'
+import { ELEMENT_KEY, SERIALIZE_PROPERTY, SERIALIZE_FLAG, ERROR_MESSAGES, PPTR_LOG_PREFIX } from './constants'
 
 const log = logger('devtools')
 
@@ -298,4 +299,32 @@ export function findByWhich (executables, priorities) {
     })
 
     return sort(uniq(installations.filter(Boolean)), priorities)
+}
+
+/**
+ * monkey patch debug package to log CDP messages from Puppeteer
+ */
+export function patchDebug (scoppedLogger) {
+    /**
+     * log puppeteer messages
+     */
+    let puppeteerDebugPkg = path.resolve(
+        path.dirname(require.resolve('puppeteer-core')),
+        'node_modules',
+        'debug')
+
+    /**
+     * check if Puppeteer has its own version of debug, if not use the
+     * one that is installed for all packages
+     */
+    if (!fs.existsSync(puppeteerDebugPkg)) {
+        puppeteerDebugPkg = require.resolve('debug')
+    }
+
+    require(puppeteerDebugPkg).log = (msg) => {
+        if (msg.includes('puppeteer:protocol')) {
+            msg = msg.slice(msg.indexOf(PPTR_LOG_PREFIX) + PPTR_LOG_PREFIX.length).trim()
+        }
+        scoppedLogger.debug(msg)
+    }
 }

--- a/packages/devtools/src/utils.js
+++ b/packages/devtools/src/utils.js
@@ -318,7 +318,12 @@ export function patchDebug (scoppedLogger) {
      * one that is installed for all packages
      */
     if (!fs.existsSync(puppeteerDebugPkg)) {
-        puppeteerDebugPkg = require.resolve('debug')
+        /**
+         * let's not get caught by our dep checker, therefor
+         * define package name in variable first
+         */
+        const pkgName = 'debug'
+        puppeteerDebugPkg = require.resolve(pkgName)
     }
 
     require(puppeteerDebugPkg).log = (msg) => {

--- a/packages/devtools/tests/__mocks__/pptrDebug.js
+++ b/packages/devtools/tests/__mocks__/pptrDebug.js
@@ -1,0 +1,2 @@
+const log = jest.fn()
+export { log }

--- a/packages/devtools/tests/utils.test.js
+++ b/packages/devtools/tests/utils.test.js
@@ -1,10 +1,15 @@
 import fs from 'fs'
+import path from 'path'
+// eslint-disable-next-line
+import { log as pptrDebugLog } from 'pptrDebug'
 import childProcess from 'child_process'
 import {
     validate, getPrototype, findElement, findElements, getStaleElementError,
     sanitizeError, transformExecuteArgs, transformExecuteResult, getPages,
-    canAccess, uniq, findByWhich
+    canAccess, uniq, findByWhich, patchDebug
 } from '../src/utils'
+
+const debug = jest.requireActual('debug')
 
 const command = {
     endpoint: '/session/:sessionId/element/:elementId/element',
@@ -50,7 +55,8 @@ jest.mock('fs', () => {
                 throw new Error('Not accessible')
             }
         }),
-        shouldThrow: (value) => (expectedThrow = value)
+        shouldThrow: (value) => (expectedThrow = value),
+        existsSync: (pkgName) => pkgName === 'pptrDebug'
     }
 })
 
@@ -65,6 +71,13 @@ jest.mock('child_process', () => {
         }),
         shouldReturn: (value) => (returnValue = value)
     }
+})
+
+jest.mock('path', () => {
+    let resolveResult = 'debug'
+    const resolve = jest.fn(() => resolveResult)
+    const setResolveResult = (result) => (resolveResult = result)
+    return { resolve, setResolveResult, dirname: jest.fn() }
 })
 
 describe('validate', () => {
@@ -364,4 +377,19 @@ test('findByWhich', () => {
     fs.shouldThrow(true)
     expect(findByWhich(['firefox'], [{ regex: /firefox/, weight: 51 }]))
         .toEqual([])
+})
+
+test('patchDebug', () => {
+    const logMock = { debug: jest.fn() }
+    patchDebug(logMock)
+    debug.log('something something - puppeteer:protocol foobar')
+    expect(logMock.debug).toBeCalledWith('foobar')
+})
+
+test('patchDebug with debug not install in puppeteer', () => {
+    const logMock = { debug: jest.fn() }
+    path.setResolveResult('pptrDebug')
+    patchDebug(logMock)
+    pptrDebugLog('something something - puppeteer:protocol barfoo')
+    expect(logMock.debug).toBeCalledWith('barfoo')
 })

--- a/packages/wdio-crossbrowsertesting-service/src/service.js
+++ b/packages/wdio-crossbrowsertesting-service/src/service.js
@@ -62,8 +62,8 @@ export default class CrossBrowserTestingService {
      * After test
      * @param {Object} test Test
      */
-    afterTest (test) {
-        if (!test.passed) {
+    afterTest (test, context, results) {
+        if (!results.passed) {
             ++this.failures
         }
     }

--- a/packages/wdio-crossbrowsertesting-service/tests/service.test.js
+++ b/packages/wdio-crossbrowsertesting-service/tests/service.test.js
@@ -148,8 +148,7 @@ describe('wdio-crossbrowsertesting-service', () => {
         const cbtService = new CrossBrowserTestingService()
         cbtService.failures = 0
 
-        cbtService.afterTest({ passed: true })
-
+        cbtService.afterTest({}, {}, { passed: true })
         expect(cbtService.failures).toEqual(0)
     })
 
@@ -157,7 +156,7 @@ describe('wdio-crossbrowsertesting-service', () => {
         const cbtService = new CrossBrowserTestingService()
         cbtService.failures = 0
 
-        cbtService.afterTest({ passed: false })
+        cbtService.afterTest({}, {}, { passed: false })
 
         expect(cbtService.failures).toEqual(1)
     })

--- a/packages/wdio-sauce-service/src/service.js
+++ b/packages/wdio-sauce-service/src/service.js
@@ -74,8 +74,8 @@ export default class SauceService {
         }
     }
 
-    afterTest (test) {
-        if (!test.passed) {
+    afterTest (test, context, results) {
+        if (!results.passed) {
             ++this.failures
         }
     }

--- a/packages/wdio-sauce-service/tests/service.test.js
+++ b/packages/wdio-sauce-service/tests/service.test.js
@@ -105,14 +105,14 @@ test('afterSuite', () => {
 
 test('afterTest', () => {
     const service = new SauceService()
-    service.beforeSession({}, {})
+    service.beforeSession({}, {}, {})
 
     expect(service.failures).toBe(0)
 
-    service.afterTest({ passed: true })
+    service.afterTest({}, {}, { passed: true })
     expect(service.failures).toBe(0)
 
-    service.afterTest({ passed: false })
+    service.afterTest({}, {}, { passed: false })
     expect(service.failures).toBe(1)
 })
 

--- a/packages/wdio-testingbot-service/src/service.js
+++ b/packages/wdio-testingbot-service/src/service.js
@@ -74,8 +74,8 @@ export default class TestingBotService {
      * After test
      * @param {Object} test Test
      */
-    afterTest (test) {
-        if (!test.passed) {
+    afterTest (test, context, results) {
+        if (!results.passed) {
             ++this.failures
         }
     }

--- a/packages/wdio-testingbot-service/tests/service.test.js
+++ b/packages/wdio-testingbot-service/tests/service.test.js
@@ -157,7 +157,7 @@ describe('wdio-testingbot-service', () => {
         const test = {
             passed: true
         }
-        tbService.afterTest(test)
+        tbService.afterTest({}, {}, test)
 
         expect(tbService.failures).toEqual(0)
     })
@@ -168,7 +168,7 @@ describe('wdio-testingbot-service', () => {
         const test = {
             passed: false
         }
-        tbService.afterTest(test)
+        tbService.afterTest({}, {}, test)
 
         expect(tbService.failures).toEqual(1)
     })

